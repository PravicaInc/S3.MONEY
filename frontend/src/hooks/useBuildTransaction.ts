import { useCurrentAccount, useSuiClient } from '@mysten/dapp-kit';
import { DryRunTransactionBlockResponse } from '@mysten/sui.js/client';
import { TransactionBlock } from '@mysten/sui.js/transactions';
import { useMutation } from '@tanstack/react-query';

export const useBuildTransaction = () => {
  const account = useCurrentAccount();
  const suiClient = useSuiClient();

  return useMutation({
    mutationFn: async (txb: TransactionBlock): Promise<DryRunTransactionBlockResponse> => {
      txb.setSenderIfNotSet(account?.address as string);

      return suiClient.dryRunTransactionBlock({
        transactionBlock: await txb.build({
          client: suiClient,
        }),
      });
    },
  });
};
