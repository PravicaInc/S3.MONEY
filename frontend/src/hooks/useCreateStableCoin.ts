import { SuiSignAndExecuteTransactionBlockOutput } from '@mysten/wallet-standard';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import { AxiosResponse } from 'axios';

import { ApiManager } from '@/services/api';

const createApi = new ApiManager(
  process.env.NEXT_PUBLIC_API_DOMAIN as string,
  '/create'
);
const cancelApi = new ApiManager(
  process.env.NEXT_PUBLIC_API_DOMAIN as string,
  '/cancel'
);
const publishedApi = new ApiManager(
  process.env.NEXT_PUBLIC_API_DOMAIN as string,
  '/published'
);

export enum suiNetworkList {
  localnet = 'localnet',
  devnet = 'devnet',
  testnet = 'testnet',
  mainnet = 'mainnet',
}

export interface CreateStableCoinData {
  walletAddress: string;
  ticker: string;
  decimals: number;
  name?: string | undefined;
  description?: string | undefined;
  maxSupply?: number;
  initialSupply?: number;
  network?: suiNetworkList;
  icon?: string;
  roles?: Record<string, string | undefined>;
}

export interface CreateStableCoinApiPostResponse {
  modules: number[][] | string[];
  dependencies: string[];
}

export interface RemoveNotPublishedStableCoinData {
  walletAddress: string;
  ticker: string;
}

export interface RemoveNotPublishedStableCoinApiPostResponse {
  status: string;
  message: string;
}

export interface SavePublishedStableCoinData {
  walletAddress: string;
  ticker: string;
  transactionID: string;
  data: SuiSignAndExecuteTransactionBlockOutput;
}

export interface SavePublishedStableCoinApiPostResponse {
  status: string;
  packages: SuiSignAndExecuteTransactionBlockOutput[];
}

export const useCreateStableCoin = () => {
  const queryClient = useQueryClient();

  return ({
    create: useMutation({
      mutationFn: async ({
        walletAddress,
        ticker,
        name,
        description,
        decimals,
        maxSupply,
        initialSupply,
        network = suiNetworkList.testnet,
        icon,
        roles,
      }: CreateStableCoinData) => createApi.post({
        address: walletAddress,
        ticker,
        name,
        description,
        decimals,
        maxSupply,
        initialSupply,
        network,
        icon_url: icon,
        roles,
      })
        .then((response: AxiosResponse<CreateStableCoinApiPostResponse>) => response.data),
    }),
    removeNotPublishedStableCoin: useMutation({
      mutationFn: async ({
        walletAddress,
        ticker,
      }: RemoveNotPublishedStableCoinData) => cancelApi.post({
        address: walletAddress,
        ticker,
        created: false,
      })
        .then((response: AxiosResponse<RemoveNotPublishedStableCoinApiPostResponse>) => response.data),
    }),
    savePublishedStableCoin: useMutation({
      mutationFn: async ({
        walletAddress,
        ticker,
        transactionID,
        data,
      }: SavePublishedStableCoinData) => publishedApi.post({
        address: walletAddress,
        ticker,
        txid: transactionID,
        created: true,
        data,
      })
        .then((response: AxiosResponse<SavePublishedStableCoinApiPostResponse>) => {
          queryClient.invalidateQueries({
            queryKey: ['stableCoinsList', walletAddress],
          });

          return response.data;
        }),
    }),
  });
};
