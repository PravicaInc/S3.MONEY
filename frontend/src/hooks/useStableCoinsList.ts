import { useQuery } from '@tanstack/react-query';
import { AxiosResponse } from 'axios';

import { ApiManager } from '@/services/api';

interface PackageApiResponse {
  deploy_addresses: {
    packageId: string,
    token_policy: string,
    token_supply: string,
    treasury_cap: string,
    token_policy_cap: string,
    pauser: string,
  },
  deploy_date: string,
  deploy_status: 'created' | 'published',
  package_name: string,
  ticker: string,
  ticker_name: string,
  txid: string,
  icon_url: string;
  package_zip: string,
  package_roles: {
    burn: string,
    freeze: string,
    cashIn: string,
    pause: string,
    cashOut: string
  },
  address_roles: string[],
}

export interface StableCoin extends Omit<PackageApiResponse, 'ticker_name' | 'icon_url'> {
  txid: string;
  ticker: string;
  name: string;
  icon?: string;
}

export interface StableCoinsListResponse {
  accountAddress: string;
  coins: StableCoin[];
}

const packagesListApi = new ApiManager(
  process.env.NEXT_PUBLIC_API_DOMAIN as string,
  '/packages'
);

export interface packagesListApiGetByKeyResponse {
  status: 'ok',
  packages: PackageApiResponse[],
}

const getStableCoinsList = async (accountAddress: string): Promise<StableCoinsListResponse> => {
  const { packages } = await packagesListApi.getByKey(accountAddress, { summary: 1 })
    .then((response: AxiosResponse<packagesListApiGetByKeyResponse>) => response.data)
    .then(data => ({
      ...data,
      packages: data.packages
        .filter(({ deploy_status }) => deploy_status === 'published')
        .sort(({ deploy_date: a }, { deploy_date: b }) => (new Date(b)).getTime() - (new Date(a)).getTime()),
    }));

  return {
    accountAddress,
    coins: packages
      .filter(({ deploy_status }) => deploy_status === 'published')
      .sort(({ deploy_date: a }, { deploy_date: b }) => (new Date(b)).getTime() - (new Date(a)).getTime())
      .map(({ txid, ticker_name, ticker, icon_url, ...otherProps }) => ({
        txid,
        ticker,
        name: ticker_name,
        icon: icon_url,
        ...otherProps,
      })),
  };
};

export const useStableCoinsList = (accountAddress: string | undefined) => {
  const query = useQuery<StableCoinsListResponse>({
    queryKey: ['stableCoinsList', accountAddress],
    queryFn: () => accountAddress
      ? getStableCoinsList(accountAddress)
      : Promise.resolve({} as StableCoinsListResponse),
  });

  return {
    ...query,
  };
};
