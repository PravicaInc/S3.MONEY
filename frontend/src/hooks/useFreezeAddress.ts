import { useSignAndExecuteTransactionBlock } from '@mysten/dapp-kit';
import { TransactionBlock } from '@mysten/sui.js/transactions';
import { useMutation } from '@tanstack/react-query';

import { useBuildTransaction } from './useBuildTransaction';

export const useFreezeAddress = () => {
  const signAndExecuteTransactionBlock = useSignAndExecuteTransactionBlock();
  const buildTransaction = useBuildTransaction();

  return useMutation({
    mutationFn: async ({
      walletAddress,
      packageName,
      packageId,
      tokenPolicyCap,
      tokenPolicy,
    }: {
      walletAddress: string,
      packageName: string,
      packageId: string,
      tokenPolicyCap: string,
      tokenPolicy: string
    }): Promise<void> => {
      const txb = new TransactionBlock();

      txb.moveCall({
        target: `${packageId}::${packageName}::freeze_address`,
        typeArguments: [`${packageId}::${packageName}::${packageName.toUpperCase()}`],
        arguments: [
          txb.object(tokenPolicyCap),
          txb.object(tokenPolicy),
          txb.pure.address(walletAddress),
        ],
      });

      await buildTransaction.mutateAsync(txb);

      await signAndExecuteTransactionBlock.mutateAsync({
        transactionBlock: txb,
        requestType: 'WaitForLocalExecution',
      });
    },
  });
};
