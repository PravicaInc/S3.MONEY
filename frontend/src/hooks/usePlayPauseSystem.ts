import { useSignAndExecuteTransactionBlock, useSuiClient } from '@mysten/dapp-kit';
import { SuiClient } from '@mysten/sui.js/client';
import { TransactionBlock } from '@mysten/sui.js/transactions';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';

import { useBuildTransaction } from './useBuildTransaction';

const is_paused = async (suiClient: SuiClient, txid: string): Promise<boolean> => {
  const { data } = (
    await suiClient.getObject({ id: txid, options: { showContent: true } })
  );

  return data?.content?.dataType === 'moveObject'
    // @ts-expect-error data.content.fields can be anything
    ? data.content.fields.value.fields.paused
    : false;
};

export const useIsSystemPaused = (txid?: string) => {
  const suiClient = useSuiClient();

  return useQuery<boolean>({
    queryKey: ['is-paused', txid],
    queryFn: () => suiClient && txid
      ? is_paused(suiClient, txid)
      : Promise.resolve(false),
  });
};

export const usePauseSystem = () => {
  const signAndExecuteTransactionBlock = useSignAndExecuteTransactionBlock();
  const queryClient = useQueryClient();
  const buildTransaction = useBuildTransaction();

  return useMutation({
    mutationFn: async ({
      pauser,
      packageName,
      packageId,
      tokenPolicyCap,
      tokenPolicy,
    }: {
      pauser: string,
      packageName: string,
      packageId: string,
      tokenPolicyCap: string,
      tokenPolicy: string
    }): Promise<void> => {
      const txb = new TransactionBlock();

      txb.moveCall({
        target: `${packageId}::${packageName}::pause`,
        typeArguments: [`${packageId}::${packageName}::${packageName.toUpperCase()}`],
        arguments: [
          txb.object(tokenPolicyCap),
          txb.object(tokenPolicy),
        ],
      });

      await buildTransaction.mutateAsync(txb);

      await signAndExecuteTransactionBlock.mutateAsync({
        transactionBlock: txb,
        requestType: 'WaitForLocalExecution',
      });

      queryClient.invalidateQueries({ queryKey: ['is-paused', pauser] });
    },
  });
};

export const usePlaySystem = () => {
  const signAndExecuteTransactionBlock = useSignAndExecuteTransactionBlock();
  const queryClient = useQueryClient();
  const buildTransaction = useBuildTransaction();

  return useMutation({
    mutationFn: async ({
      pauser,
      packageName,
      packageId,
      tokenPolicyCap,
      tokenPolicy,
    }: {
      pauser: string,
      packageName: string,
      packageId: string,
      tokenPolicyCap: string,
      tokenPolicy: string
    }): Promise<void> => {
      const txb = new TransactionBlock();

      txb.moveCall({
        target: `${packageId}::${packageName}::unpause`,
        typeArguments: [`${packageId}::${packageName}::${packageName.toUpperCase()}`],
        arguments: [
          txb.object(tokenPolicyCap),
          txb.object(tokenPolicy),
        ],
      });

      await buildTransaction.mutateAsync(txb);

      await signAndExecuteTransactionBlock.mutateAsync({
        transactionBlock: txb,
        requestType: 'WaitForLocalExecution',
      });

      queryClient.invalidateQueries({ queryKey: ['is-paused', pauser] });
    },
  });
};
