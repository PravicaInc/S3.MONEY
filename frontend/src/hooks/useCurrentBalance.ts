import { useCurrentAccount, useSuiClient } from '@mysten/dapp-kit';
import { GetOwnedObjectsParams, SuiClient } from '@mysten/sui.js/client';
import { UndefinedInitialDataOptions, useQuery } from '@tanstack/react-query';

import convertSuiBalanceStringToNumber from '@/utils/convertSuiBalanceStringToNumber';

import { StableCoin } from './useStableCoinsList';

export interface StableCoinBalanceObject {
  data: {
    objectId: string,
    digest: string,
    content: {
      dataType: 'moveObject',
      type: string,
      hasPublicTransfer: boolean,
      fields: {
        balance: string | number,
        id: {
          id: string,
        },
      },
    },
  },
}

const getCurrentBalance = async (suiClient: SuiClient, accountAddress: string): Promise<number> => {
  const { totalBalance } = await suiClient.getBalance({
    owner: accountAddress,
  });

  return convertSuiBalanceStringToNumber(totalBalance);
};

export const useCurrentSuiBalance = (
  queryOption?: Omit<UndefinedInitialDataOptions<number, Error, number>, 'queryKey'>
) => {
  const account = useCurrentAccount();
  const suiClient = useSuiClient();

  return useQuery<number>({
    ...queryOption,
    queryKey: ['current-sui-balance', account?.address],
    queryFn: () => suiClient && account?.address
      ? getCurrentBalance(suiClient, account?.address)
      : Promise.resolve(0),
  });
};

export const useCurrentStableCoinBalance = (walletAddress?: string, stableCoin?: StableCoin) => {
  const suiClient = useSuiClient();

  return useQuery<number>({
    queryKey: ['current-stable-coin-balance', walletAddress, stableCoin?.ticker],
    queryFn: async () => (
      walletAddress && stableCoin
        ? (await getAllOwnedObjects(suiClient, { owner: walletAddress }))
          .filter(
            obj => obj.data?.content?.dataType === 'moveObject'
              && obj.data?.content?.type == `0x2::token::Token<${stableCoin.deploy_addresses.packageId}::${stableCoin.package_name}::${stableCoin.package_name.toUpperCase()}>`
          )
          .map(coin => parseFloat(coin.data.content.fields.balance as string))
          .reduce((sum, next) => sum + next)
        : Promise.resolve(0)
    ),
  });
};

export async function getAllOwnedObjects(
  suiClient: SuiClient,
  {
    owner,
    cursor,
  }: GetOwnedObjectsParams
): Promise<StableCoinBalanceObject[]> {
  const { data, hasNextPage, nextCursor } = await suiClient.getOwnedObjects({
    owner,
    cursor,
    options: {
      showContent: true,
    },
  });

  return [
    ...data as unknown as StableCoinBalanceObject[],
    ...[
      hasNextPage && nextCursor
        ? await getAllOwnedObjects(suiClient, { owner, cursor: nextCursor })
        : [],
    ],
  ].flat();
}
