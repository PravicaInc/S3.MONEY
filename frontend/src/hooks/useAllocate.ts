import { useSignAndExecuteTransactionBlock, useSuiClient } from '@mysten/dapp-kit';
import { TransactionBlock } from '@mysten/sui.js/transactions';
import { useMutation, useQueryClient } from '@tanstack/react-query';

import { useBuildTransaction } from './useBuildTransaction';
import { getAllOwnedObjects } from './useCurrentBalance';
import { joinCoins } from './useStableCoinSupply';

export const useAllocate = () => {
  const signAndExecuteTransactionBlock = useSignAndExecuteTransactionBlock();
  const buildTransaction = useBuildTransaction();
  const queryClient = useQueryClient();
  const suiClient = useSuiClient();

  return useMutation({
    mutationFn: async ({
      senderAddresses,
      recipientAddresses,
      packageName,
      packageId,
      treasuryCap,
      tokenPolicy,
      tokenSupply,
      amount,
    }: {
      senderAddresses: string,
      recipientAddresses: string,
      packageName: string,
      packageId: string,
      treasuryCap: string,
      tokenPolicy: string,
      tokenSupply: string,
      amount: number,
    }): Promise<void> => {
      const coins = (await getAllOwnedObjects(suiClient, { owner: senderAddresses }))
        .filter(
          obj => obj.data?.content?.dataType === 'moveObject'
            && obj.data?.content?.type == `0x2::token::Token<${packageId}::${packageName}::${packageName.toUpperCase()}>`
        )
        .map(coin => {
          coin.data.content.fields.balance = parseFloat(coin.data.content.fields.balance as string);

          return coin;
        })
        .sort((coinA, coinB) => (
          (coinB.data.content.fields.balance as number) - (coinA.data.content.fields.balance as number)
        ));

      const coinsForAllocate = [];

      for (let currentAmount = amount, i = 0; currentAmount > 0 && i < coins.length; i++) {
        coinsForAllocate.push(coins[i]);

        currentAmount -= (coins[i].data?.content.fields.balance as number);
      }

      const txb = new TransactionBlock();

      const { newCoin, newBalance } = joinCoins(
        txb,
        coinsForAllocate,
        packageId,
        packageName
      );

      const coinForAllocate = amount === newBalance
        ? newCoin
        : txb.moveCall({
          target: '0x2::token::split',
          typeArguments: [`${packageId}::${packageName}::${packageName.toUpperCase()}`],
          arguments: [
            txb.object(newCoin),
            txb.pure.u64(amount),
          ],
        });

      txb.moveCall({
        target: `${packageId}::${packageName}::allocate`,
        typeArguments: [`${packageId}::${packageName}::${packageName.toUpperCase()}`],
        arguments: [
          txb.object(treasuryCap),
          txb.object(tokenPolicy),
          txb.object(coinForAllocate),
          txb.pure.address(recipientAddresses),
        ],
      });

      await buildTransaction.mutateAsync(txb);

      await signAndExecuteTransactionBlock.mutateAsync({
        transactionBlock: txb,
        requestType: 'WaitForLocalExecution',
      });

      queryClient.invalidateQueries({
        queryKey: ['stable-coin-current-supply', treasuryCap],
      });
      queryClient.invalidateQueries({
        queryKey: ['stable-coin-max-supply', tokenSupply],
      });
      queryClient.invalidateQueries({
        queryKey: ['current-stable-coins-balances'],
      });
      queryClient.invalidateQueries({
        queryKey: ['current-stable-coin-balance'],
      });
    },
  });
};
