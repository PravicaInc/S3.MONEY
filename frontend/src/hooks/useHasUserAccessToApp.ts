import { useQuery, UseQueryOptions } from '@tanstack/react-query';

import { useCurrentStableCoinBalance } from './useCurrentBalance';
import { StableCoin } from './useStableCoinsList';

export const useHasUserAccessToApp = (
  walletAddress?: string,
  queryOption?: Omit<UseQueryOptions<boolean, Error, boolean>, 'queryKey'>
) => {
  const hasAccessStablecoin = !!(
    process.env.NEXT_PUBLIC_ACCESS_STABLECOIN_PACKAGE_ID
      && process.env.NEXT_PUBLIC_ACCESS_STABLECOIN_PACKAGE_NAME
      && process.env.NEXT_PUBLIC_ACCESS_STABLECOIN_TICKER
  );
  const stableCoinBalance = useCurrentStableCoinBalance(
    walletAddress,
    {
      deploy_addresses: {
        packageId: process.env.NEXT_PUBLIC_ACCESS_STABLECOIN_PACKAGE_ID,
      },
      package_name: process.env.NEXT_PUBLIC_ACCESS_STABLECOIN_PACKAGE_NAME,
      ticker: process.env.NEXT_PUBLIC_ACCESS_STABLECOIN_TICKER,
    } as StableCoin
  );

  return useQuery<boolean>({
    ...queryOption,
    queryKey: ['has-user-access-to-app', walletAddress, stableCoinBalance.data],
    queryFn: async () => (
      hasAccessStablecoin
        ? (
          walletAddress
            ? !!stableCoinBalance.data
            : Promise.resolve(false)
        )
        : Promise.resolve(true)
    ),
    enabled: hasAccessStablecoin
      ? stableCoinBalance.data !== undefined
      : true,
  });
};
