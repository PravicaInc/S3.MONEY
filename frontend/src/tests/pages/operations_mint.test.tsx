import { useAutoConnectWallet, useCurrentAccount } from '@mysten/dapp-kit';
import { cleanup, RenderResult } from '@testing-library/react';
import { ReadonlyURLSearchParams, usePathname, useSearchParams } from 'next/navigation';
import qs from 'qs';

import '@testing-library/jest-dom';

import OperationsMintPage from '@/app/dashboard/operations/mint/page';

import { useStableCoinsList } from '@/hooks/useStableCoinsList';

import { createMockStableCoinList, TEST_STABLE_COIN1 } from '@/tests/utils/create_mock_stable_coin_list';
import { createMockAccount } from '@/tests/utils/create_mock_wallet_account';
import { renderWithProviders } from '@/tests/utils/render_with_providers';

const renderSignInPageWithProviders = () => renderWithProviders(<OperationsMintPage />);

let renderResult: RenderResult;

describe('Operations/Mint page:', () => {
  afterAll(cleanup);

  describe('Operations/Mint page with connected state:', () => {
    beforeEach(() => {
      const domain = 'https://localhost:3001';
      const url = new URL(`${domain}/dashboard/operations/mint?${qs.stringify({
        txid: TEST_STABLE_COIN1.txid,
      })}`);

      if (renderResult?.unmount) {
        renderResult.unmount();
      }

      jest.mocked(usePathname).mockImplementation(() => url.pathname);
      jest.mocked(useSearchParams).mockImplementation(() => (
        new URLSearchParams(url.search)
      ) as ReadonlyURLSearchParams);
      jest.mocked(useAutoConnectWallet).mockImplementation(() => 'attempted');
      jest.mocked(useCurrentAccount).mockImplementation(createMockAccount);
      jest.mocked(useStableCoinsList).mockImplementation(
        createMockStableCoinList()
      );

      renderResult = renderSignInPageWithProviders();
    });

    it('Snapshot', () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });
});
