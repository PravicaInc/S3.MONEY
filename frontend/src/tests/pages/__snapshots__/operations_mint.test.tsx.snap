// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operations/Mint page: Operations/Mint page with connected state: Snapshot 1`] = `
<div>
  <style>
    [data-dapp-kit] {--dapp-kit-blurs-modalOverlay:blur(0);--dapp-kit-backgroundColors-primaryButton:#F6F7F9;--dapp-kit-backgroundColors-primaryButtonHover:#F0F2F5;--dapp-kit-backgroundColors-outlineButtonHover:#F4F4F5;--dapp-kit-backgroundColors-modalOverlay:rgba(24 36 53 / 20%);--dapp-kit-backgroundColors-modalPrimary:white;--dapp-kit-backgroundColors-modalSecondary:#F7F8F8;--dapp-kit-backgroundColors-iconButton:transparent;--dapp-kit-backgroundColors-iconButtonHover:#F0F1F2;--dapp-kit-backgroundColors-dropdownMenu:#FFFFFF;--dapp-kit-backgroundColors-dropdownMenuSeparator:#F3F6F8;--dapp-kit-backgroundColors-walletItemSelected:white;--dapp-kit-backgroundColors-walletItemHover:#3C424226;--dapp-kit-borderColors-outlineButton:#E4E4E7;--dapp-kit-colors-primaryButton:#373737;--dapp-kit-colors-outlineButton:#373737;--dapp-kit-colors-iconButton:#000000;--dapp-kit-colors-body:#182435;--dapp-kit-colors-bodyMuted:#767A81;--dapp-kit-colors-bodyDanger:#FF794B;--dapp-kit-radii-small:6px;--dapp-kit-radii-medium:8px;--dapp-kit-radii-large:12px;--dapp-kit-radii-xlarge:16px;--dapp-kit-shadows-primaryButton:0px 4px 12px rgba(0, 0, 0, 0.1);--dapp-kit-shadows-walletItemSelected:0px 2px 6px rgba(0, 0, 0, 0.05);--dapp-kit-fontWeights-normal:400;--dapp-kit-fontWeights-medium:500;--dapp-kit-fontWeights-bold:600;--dapp-kit-fontSizes-small:14px;--dapp-kit-fontSizes-medium:16px;--dapp-kit-fontSizes-large:18px;--dapp-kit-fontSizes-xlarge:20px;--dapp-kit-typography-fontFamily:ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";--dapp-kit-typography-fontStyle:normal;--dapp-kit-typography-lineHeight:1.3;--dapp-kit-typography-letterSpacing:1;}
  </style>
  <div
    class="max-w-screen-2xl mx-auto p-8 flex items-center justify-center h-full"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-spinner animate-spin text-actionPrimary h-8"
      data-icon="spinner"
      data-prefix="fas"
      data-testid="loader"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zm0 416a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zM48 304a48 48 0 1 0 0-96 48 48 0 1 0 0 96zm464-48a48 48 0 1 0 -96 0 48 48 0 1 0 96 0zM142.9 437A48 48 0 1 0 75 369.1 48 48 0 1 0 142.9 437zm0-294.2A48 48 0 1 0 75 75a48 48 0 1 0 67.9 67.9zM369.1 437A48 48 0 1 0 437 369.1 48 48 0 1 0 369.1 437z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
